app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/chart/BeaconLineChart.java

```diff
@@ -317,72 +323,78 @@ protected void drawBeacons(Canvas canvas) {
 
     @Override
     protected void drawBeacon(Canvas canvas, Beacon beacon) {
-        currentBeaconIndex = beacons.indexOf(beacon);
-
-        linePaint.setShader(createLineShader(getBeaconColor(beacon, coloringMode, currentBeaconIndex)));
-        linePaint.setAlpha(128);
-
-        lastLinePoint = null;
-        lastAdvertisingPacket = null;
-
+        prepareDraw(beacon);
         for (AdvertisingPacket advertisingPacket : (List<AdvertisingPacket>) beacon.getAdvertisingPackets()) {
             if (advertisingPacket.getTimestamp() < minimumAdvertisingTimestamp) {
                 continue;
             }
             currentLinePoint = getPointFromAdvertisingPacket(beacon, advertisingPacket, currentLinePoint);
-            canvas.drawCircle(currentLinePoint.x, currentLinePoint.y, pixelsPerDip * 1.5f, linePaint);
-
-            if (lastLinePoint != null && lastAdvertisingPacket != null) {
-                if (advertisingPacket.getTimestamp() < lastAdvertisingPacket.getTimestamp() + 5000) {
-                    canvas.drawLine(
-                            lastLinePoint.x,
-                            lastLinePoint.y,
-                            currentLinePoint.x,
-                            currentLinePoint.y,
-                            linePaint
-                    );
-                }
-            } else {
-                lastLinePoint = new PointF();
-            }
+            drawNextPoint(canvas, beacon, advertisingPacket);
+        }
+        currentBeaconIndex++;
+        fadeLastPoint(canvas);
+    }
 
-            lastLinePoint.x = currentLinePoint.x;
-            lastLinePoint.y = currentLinePoint.y;
-            lastAdvertisingPacket = advertisingPacket;
-            currentBeaconIndex++;
+    protected void prepareDraw(Beacon beacon) {
+        lastLinePoint = null;
+        lastAdvertisingPacket = null;
+
+        linePaint.setShader(createLineShader(getBeaconColor(beacon, coloringMode, currentBeaconIndex)));
+        linePaint.setAlpha(128);
+    }
+
+    protected void drawNextPoint(Canvas canvas, Beacon beacon, AdvertisingPacket advertisingPacket) {
+        canvas.drawCircle(currentLinePoint.x, currentLinePoint.y, pixelsPerDip * 1.5f, linePaint);
```

Comment:
`canvas.drawCircle(...)` has nothing to do with `drawNextPoint()`. It is a side-effect and should be moved.